### PR TITLE
[5.0] Moved From Boris To Psysh

### PR DESCRIPTION
**Pull request by GrahamCampbell**
Replaces #6684.

This moves us from boris to psysh. Psysh is better than boris, and more importantly, supports windows, and doesn't have a hard requirement on certain extensions that would cause the framework to be unusable because composer with cry.

This pull also upgrades us from super closure ~1.0 to ~2.0, and from class preloader 1.0.x to ~1.2.

NB, under the hood, all these changes mean that we're moving from php parser 0.9 to ~1.0.

### CODE REVIEW

**Comment by GrahamCampbell** [comment]
Related: https://github.com/jeremeamia/super_closure/issues/34.

**Comment by joecohens** [comment]
:+1:

**Comment by GrahamCampbell** [comment]
Failing tests are due the fact that packagist is totally retarded when it comes to dealing with deleted branches. @jeremeamia should be manually fixing this shortly.

**Comment by GrahamCampbell** [comment]
And we're good to go now. Everything is green!

File: src/Illuminate/Foundation/Console/TinkerCommand.php [hunk 1]
```diff
@@ -26,26 +26,11 @@ class TinkerCommand extends Command {
...
-	 * Determine if the current environment supports Boris.
-	 *
-	 * @return bool
-	 */
-	protected function supportsBoris()
-	{
-		return extension_loaded('readline') && extension_loaded('posix') && extension_loaded('pcntl');
+		$shell = new Shell();
+
+		$shell->run();
```
Comment (JosephSilber) [inline src/Illuminate/Foundation/Console/TinkerCommand.php]: Why not just `(new Shell)->run()`?

Comment (GrahamCampbell) [inline src/Illuminate/Foundation/Console/TinkerCommand.php]: I prefer defining the variable first, but I can change it if you really want @taylorotwell.

**Comment by gmsantosxl** [comment]
:+1: for windows support

**Comment by GrahamCampbell** [comment]
Yeh. This was a big problem before. composer.lock files generated on linux would mean that the dependencies wouldn't even install on windows.

**Comment by GrahamCampbell** [comment]
Ping @taylorotwell. All dependencies are stable now.

**Comment by JosephSilber** [comment]
@GrahamCampbell maybe do a rebase?

**Comment by GrahamCampbell** [comment]
I'm not at my PC atm. It might be faster for Taylor to rebase and merge.

**Comment by GrahamCampbell** [comment]
Done.

**Comment by joecohens** [comment]
:smile: :smile: :smile:

**Comment by jeremeamia** [comment]
:smile:

**Comment by GrahamCampbell** [comment]
//cc @bobthecow Laravel 5's now shipping with psysh.

File: src/Illuminate/Foundation/Console/TinkerCommand.php (hunk 1 above)
Comment (bobthecow) [inline src/Illuminate/Foundation/Console/TinkerCommand.php]: Another reason would be the context variables and awesome custom presenters and commands you're going to inject here :)

([See the equivalent bit of code in ezsh](https://github.com/lolautruche/ezsh/blob/master/bin/ezsh#L135-L165))

Comment (GrahamCampbell) [inline src/Illuminate/Foundation/Console/TinkerCommand.php]: I'll change this then. :)

**Comment by bobthecow** [comment]
@GrahamCampbell Nice!

**Comment by webcraft** [comment]
It seems this PR has introduced some trouble in Mail::queue(). When trying to pass a closure to Mail::queue(), you'll get an InvalidArgumentException in Mailer.php line 393: Callback is not valid.

E.g. try this on a fresh install (latest L5 commit):

```php
Route::get('/queuetest', function () {
	Mail::queue('emails.password', ['token' => '1234'], function ($message) {
		$message->from('test@example.com', 'test');
		$message->to('test@example.com', 'test')->subject('Queue test)');
	});
});
```

If we could change the line
https://github.com/laravel/framework/blob/master/src/Illuminate/Mail/Mailer.php#L384

from 
```php
	if ($callback instanceof Closure)
```
to
```php
	if ($callback instanceof SerializableClosure)
```
things would work out normal.
Or will this introduce other issues?

Anyway, thanks @GrahamCampbell for the great effort!

**Comment by GrahamCampbell** [comment]
Thanks for the catch. I'll fix this.

**Comment by GrahamCampbell** [comment]
See #7167.

**Pull request merged**